cmd/limactl/shell.go

```diff
@@ -178,7 +180,25 @@ func shellAction(cmd *cobra.Command, args []string) error {
 	} else {
 		shell = shellescape.Quote(shell)
 	}
-	script := fmt.Sprintf("%s ; exec %s --login", changeDirCmd, shell)
+	// Handle environment variable propagation
+	var envPrefix string
+	withEnv, err := cmd.Flags().GetBool("preserve-env")
+	if err != nil {
+		return err
+	}
+	if withEnv {
+		config := envutil.GetFilterConfig()
+		filteredEnv := envutil.FilterEnvironment(config)
```

Comment:
You probably split this into 2 functions for testability, but I think for actual usage there should be no need to pass the `config`.

```suggestion
		filteredEnv := envutil.FilterEnvironment()
```

You can still have a separate function for testing, and `FilterEnvironment` just becomes a wrapper that calls `GetFilterConfig` internally.